src/main/java/org/spongepowered/api/resource/ResourcePath.java

```diff
@@ -0,0 +1,254 @@
+ *
+ * <p>A resource path should be usable in a {@link Map}, so implementations
+ * should override {@link #hashCode()} and {@link #equals(Object)}.</p>
+ */
+public interface ResourcePath extends Comparable<ResourcePath>, Iterable<ResourcePath> {
+
+    /**
+     * Creates a new {@link Builder} to build a {@link ResourcePath}.
+     *
+     * @return The new builder
+     */
+    static Builder builder() {
+        return Sponge.getRegistry().createBuilder(Builder.class);
+    }
+
+    /**
+     * Creates a new {@link ResourcePath} using the given namespace and path.
+     *
+     * @param namespace The namespace to use
+     * @param path The path to use
+     * @return A new ResourcePath
+     */
+    static ResourcePath of(String namespace, String path) throws IllegalArgumentException {
+        return builder().namespace(namespace).path(path).build();
+    }
+
+    /**
+     * Parses a new {@link ResourcePath} from the given string representation.
+     * The representation should include both the namespace and path, separated
+     * by a colon (:). Path items should be separated by a forward slash (/).
+     * If the namespace is excluded, the default of minecraft is used instead.
+     * <p>
+     * <p>e.g. {@code foo:bar/custom.json} points to the
+     * {@code bar/custom.json} file in the {@code foo} namespace.</p>
+     *
+     * @param path The path
+     * @return A new ResourcePath
+     */
+    static ResourcePath parse(String path) throws IllegalArgumentException {
+        return builder().parse(path).build();
+    }
+
+    /**
+     * Gets the namespace portion of this resource path.
+     *
+     * @return The namespace
+     */
+    String getNamespace();
+
+    /**
+     * Gets the path portion of this resource path
+     *
+     * @return The path
+     */
+    String getPath();
+
+    /**
+     * Gets the root path from this path.
+     *
+     * @return The root path
+     */
+    ResourcePath getRoot();
+
+    /**
+     * Gets the immediate parent for this {@link ResourcePath}. If this path is
+     * the root path, itself is returned.
+     *
+     * @return The parent path
+     * @see #getRoot()
+     */
+    ResourcePath getParent();
+
+    /**
+     * Gets a list of names which make up this {@link ResourcePath}. Order of
+     * names will be root first, with the last name being this path's
+     * filename.
+     *
+     * @return The list of paths
+     */
+    List<String> getNames();
+
+    /**
+     * Tests if this {@link ResourcePath} is a direct child of the given other path.
+     *
+     * @param parent The other resource path to check against
+     * @return True if this is a child to other
+     */
+    boolean isParent(ResourcePath parent);
+
+    /**
+     * Gets the file name portion of this {@link ResourcePath}.
+     *
+     * @return The file name of the path
+     * @see #getBaseName()
+     * @see #getExtension()
+     */
+    String getFileName();
```

Comment:
I wouldn't call this `getFileName` since there is no guarantee that it's actually a file, it can also be a directory. This also breaks the `getBaseName` method.

Reply:
I disagree. Yes, there is no guarantee that it is a file, but that's not up to the path to decide. The path has no say in whether something exists, or is a file (or directory).

And `getBaseName` was named after `FilenameUtils.getBaseName`, which I added because IOUtils is not included in the API.